### PR TITLE
Fix dockerfile & entrypoint:

### LINKED ISSUE #1581: Problem with Docker image and saving identity
I am having trouble running the official Docker image. The first run is fine but subsequent runs that re-use the zerotier db don't startup correctly. I think this is an entrypoint.sh.release bug.

# Repro steps

1. First run:
```
docker volume create zerotier-db
docker run --name myzerotier --rm --cap-add NET_ADMIN --device /dev/net/tun -v zerotier-db:/var/lib/zerotier-one zerotier/zerotier:1.8.4 <REDACTED-ZEROTIER-NETWORK-ID>
```
Output:
```
starting zerotier
cat: /var/lib/zerotier-one/zerotier-one.pid: No such file or directory
cat: /var/lib/zerotier-one/zerotier-one.pid: No such file or directory
joining networks: 1c33c1ced0d58dd6
joining 1c33c1ced0d58dd6
0 join connection failed
joining 1c33c1ced0d58dd6 failed; trying again in 1s
200 join OK
```
2. Validate network is awaiting authorization:
```
$ docker exec `docker container ls | grep zero | awk '{ print $1 }'` zerotier-cli listnetworks
200 listnetworks <nwid> <name> <mac> <status> <type> <dev> <ZT assigned ips>
200 listnetworks <REDACTED> ACCESS_DENIED PRIVATE <REDACTED>
```
3. Authorize node in ZeroTier Central
4. Verify network is attached:
```
$ docker exec `docker container ls | grep zero | awk '{ print $1 }'` zerotier-cli listnetworks
200 listnetworks <nwid> <name> <mac> <status> <type> <dev> <ZT assigned ips>
200 listnetworks <REDACTED> OK PRIVATE <REDACTED><REDACTED>/16
```
5. Kill container (`ctrl-c` doesn't work, see #1580):
```
docker kill `docker container ls | grep zero | awk '{ print $1 }'`
```
6. Run container again:
```
docker run --name myzerotier --rm --cap-add NET_ADMIN --device /dev/net/tun -v zerotier-db:/var/lib/zerotier-one zerotier/zerotier:1.8.4 <REDACTED-ZEROTIER-NETWORK-ID>
```
Output:
```
starting zerotier
zerotier hasn't started, waiting a second
zerotier hasn't started, waiting a second
zerotier hasn't started, waiting a second
zerotier hasn't started, waiting a second
...
```
This never completes. The last line is printed over and over. I would expect subsequent runs to re-use the db in the named docker volume `zerotier-db` but it looks like it's failing to startup. Obviously, being able to make the identity and secret stable in a named volume so subsequent runs can keep their identity and authorization is critical to running in a Docker container.
Interestingly, this might actually be an issue with entrypoint.sh.release's logic, because everything looks fine in the container when this is happening:
```
$ docker exec -it `docker container ls | grep zero | awk '{ print $1 }'` /bin/bash
$ zerotier-cli info
200 info <REDACTED> 1.8.4 ONLINE
$ zerotier-cli listnetworks
200 listnetworks <nwid> <name> <mac> <status> <type> <dev> <ZT assigned ips>
200 listnetworks <REDACTED> OK PRIVATE <REDACTED><REDACTED>/16
$ cat /var/lib/zerotier-one/zerotier-one.pid
10
...
```

### PR DESCRIPTION
**Pull request by erikh**
- Resolve issue with join not being checked properly for success without
  using external tools
- Resolve issue where initial boot was not being checked properly
- Now output errors when zerotier fails to start

closes #1581

cc @altano for inspiration for this patch

Signed-off-by: Erik Hollensbe <git@hollensbe.org>

### CODE REVIEW

File: entrypoint.sh.release [hunk 1]
```diff
@@ -55,9 +56,11 @@ for i in "$@"
 do
   echo "joining $i"
 
-  while ! zerotier-cli join "$i"
+  zerotier-cli join "$i"
+
+  while [ "$(zerotier-cli get $i status)" != "OK" ]
```
Comment (erikh) [inline entrypoint.sh.release:61]: @altano right here is your healthcheck...

**Comment by erikh** [comment]
We'll release a new image based on 1.8.x based on these changes once everyone agrees it's a good idea.

**Comment by erikh** [comment]
I'll look into it; thanks

------- Original Message -------

**Comment by erikh** [comment]
Waiting for 1.8.7 to land before I work further on this.

**Comment by erikh** [comment]
```
% docker run --name myzerotier --rm --cap-add NET_ADMIN --device /dev/net/tun zerotier/zerotier:1.8.6 <network id>

starting zerotier
cat: /var/lib/zerotier-one/zerotier-one.pid: No such file or directory
cat: /var/lib/zerotier-one/zerotier-one.pid: No such file or directory
zerotier hasn't started, waiting a second
tail: cannot open 'nohup.out' for reading: No such file or directory
joining networks: <network id>
joining <network id>
200 join OK
joining <network id> failed (are they added in central?); trying again in 1s
joining <network id> failed (are they added in central?); trying again in 1s
joining <network id> failed (are they added in central?); trying again in 1s
joining <network id> failed (are they added in central?); trying again in 1s
joining <network id> failed (are they added in central?); trying again in 1s
^Cjoining <network id> failed (are they added in central?); trying again in 1s
Killing zerotier
```

I have no idea why ^C isn't working. what platforms are you on? This is probably docker and not us at this point.

**Comment by erikh** [comment]
also this is the build in this PR, not the image on dockerhub.

**Comment by erikh** [comment]
Now I'm able to reproduce the ^C issue. Looking into it.

**Comment by erikh** [comment]
^C issue should be in the latest patch; also I added a HEALTHCHECK. @altano would you mind testing this image for me if you're not committed to your own implementation? Would be nice to find a happy medium we can rely on for everyone.

**Comment by erikh** [comment]
an example of the healthcheck in action, fwiw;

```
[43] erikh@islay zerotier/ZeroTierOne% docker ps                                                                               [18:12]
CONTAINER ID   IMAGE                     COMMAND                  CREATED              STATUS                          PORTS     NAMES
06ccb698247d   zerotier/zerotier:1.8.6   "/entrypoint.sh 8304…"   About a minute ago   Up About a minute (unhealthy)             myzerotier
```

**Comment by altano** [comment]
> @altano would you mind testing this image for me if you're not committed to your own implementation? Would be nice to find a happy medium we can rely on for everyone.

I am not at all committed to my implementation and would be happy to test and switch over to this official one. I’ll probably have time to take a look this weekend. Thanks for working on this!

**Comment by erikh** [comment]
Ok, cool! I think once @glimberg has verified the tickets can be closed we'll merge this so we can push images. Then we'll address testing results after the fact if necessary with a second possible push; I'd like this out there sooner rather than later as I imagine it's pretty broken in its current state.

**Comment by glimberg** [comment]
Good news is that it appears to consistently join the networks now.   :partying_face:  Can ping the container from another machine on the lan.  All that good stuff.  Hooray!

This error/warning message pops up sometimes
```
joining 15707c8737d49dab failed (are they added in central?); trying again in 1s
```

Everything still works after that, but it's a little disconcerting to see

Finally, our good friend CTRL-C.  Still no joy without `-i -t`.   `docker stop $contianer`  works fine as always. I'm honestly not too bothered about the CTRL-C thing.  It seems to be a common thing out on the google.  That's how I found the `-i -t` workaround to begin with :man_shrugging:

**Comment by erikh** [comment]
> Good news is that it appears to consistently join the networks now. partying_face Can ping the container from another machine on the lan. All that good stuff. Hooray!
> 
> This error/warning message pops up sometimes
> 
> ```
> joining 15707c8737d49dab failed (are they added in central?); trying again in 1s
> ```
> 
> Everything still works after that, but it's a little disconcerting to see

So on this, the only thing I could do is sleep a little longer, which I'm sure I don't have to explain won't get rid of it entirely. It's basically going to be printed while the zerotier-one instance is waiting to join the network; I'll adjust the error message a little bit.

**Comment by glimberg** [comment]
> Oh hm. Make sure you built against the latest patches; this error message isn't in the latest one. :)

Oh ok.  I tried the one you posted in Mattermost last night

**Comment by erikh** [comment]
yeah, the one in git works much better; ctrl+c works too.

**Comment by glimberg** [comment]
Oh sweet!  Yeah that's working great now.  Even ctrl-c

**Comment by erikh** [comment]
:shipit: 

Gonna merge with your permission

**Comment by erikh** [comment]
zerotier/zerotier:1.8.6 was pushed to dockerhub as a result of this change

**Comment by altano** [comment]
Sorry, @erikh, that it took me so long to get to this. I just tested this out and everything looks great:

✅ ctrl-c works
✅ healthcheck works
✅ I can successfully join a network

I tested with this docker-compose.yaml:
```yaml
version: "3.7"

volumes:
  data:

services:
  zerotier:
    image: zerotier/zerotier:1.8.6
    cap_add:
      - NET_ADMIN
    devices:
      - /dev/net/tun
    volumes:
      - data:/var/lib/zerotier-one
    command: <zt-network-address>

  whoami:
    image: containous/whoami:latest
    depends_on:
      - zerotier
    network_mode: service:zerotier
    restart: unless-stopped
```

I added a note to the top of my Docker image github repo that no one should use it. Thanks for the quick turnaround on these improvements!

If you have time, I have one small feature request: can you dump the ZT address to the Docker logs? That way, when I first bring up the container stack, I know I'm approving the join request for the right node. I can get the address, given the stack defined above, with this:
```
$ docker-compose exec zerotier zerotier-cli info
200 info <redacted-address> 1.8.6 ONLINE
```
But dumping it to the logs would make things easier.

Thanks again!

**Comment by erikh** [comment]
Thanks for testing, and sure. Mind creating a new ticket for that?

**Comment by altano** [comment]
> Thanks for testing, and sure. Mind creating a new ticket for that?

Here's a PR: https://github.com/zerotier/ZeroTierOne/pull/1619

**Pull request merged**